File: files/200-299/241/CH-241.R
library(tidyverse)
library(readxl)
library(slider)

path = "files/200-299/241/CH-241 Moving Average.xlsx"
input = read_excel(path, range = "B2:C20")
test = read_excel(path, range = "G2:G20") %>%
  mutate(`Moving Average` = as.numeric(`Moving Average`))

result = input %>%
  mutate(
    `Moving Average` = slide_dbl(
      .x = Sales,
      .f = function(window_values) {
        non_zero_values <- window_values[window_values != 0]
        if (length(non_zero_values) >= 2) {
          mean(tail(non_zero_values, 2))
        } else {
          NA_real_
        }
      },
      .before = Inf,
      .after = -1,
      .complete = TRUE
    )
  )

all.equal(result$`Moving Average`, test$`Moving Average`, check.attributes = FALSE) # TRUE

File: files/200-299/241/CH-241.py
import pandas as pd

path = "200-299/241/CH-241 Moving Average.xlsx"
input = pd.read_excel(path, usecols="B:C", skiprows=1, nrows=18)
test = pd.read_excel(path, usecols="G", skiprows=1, nrows=18)
test['Moving Average'] = pd.to_numeric(test['Moving Average'], errors='coerce')

def moving_average(sales):
    result = []
    for i in range(len(sales)):
        non_zero = [x for x in sales[:i] if x != 0]
        if len(non_zero) >= 2:
            result.append(sum(non_zero[-2:]) / 2)
        else:
            result.append(float('nan'))
    return result

input['Moving Average'] = moving_average(input['Sales'])

print(input['Moving Average'].equals(test['Moving Average']))
